File: leetcode/src/main/java/com/thanos/algorithm/golang/add_two_numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{0, nil}
    cur := head
    c1 := l1
    c2 := l2

    carry := 0
    sum := 0
    for ; c1 != nil || c2 != nil || carry != 0; {
        sum = carry;
        if c1 != nil {
            sum = sum + c1.Val
            c1 = c1.Next
        }
        if c2 != nil {
            sum = sum + c2.Val
            c2 = c2.Next
        }
        carry = sum / 10
        cur.Next = &ListNode{sum % 10, nil}
        cur = cur.Next
    }
    return head.Next
}

File: leetcode/src/main/ruby/add_two_numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    head = ListNode.new(0)
    cur, c1, c2 = head, l1, l2
    carry, sum = 0, 0

    while c1 || c2 || carry != 0
        sum = carry
        if c1 then
            sum += c1.val
            c1 = c1.next
        end
        if c2 then
            sum += c2.val
            c2 = c2.next
        end

        carry, digit = sum.divmod(10)
        cur.next = ListNode.new(digit)
        cur = cur.next
    end

    head.next
end

